File: deep-learning/artificial-nerural-network/ANN_RC.R
#Artificial Neural network 


# Importing the dataset
dataset = read.csv('Churn_Modelling.csv')
dataset = dataset[4:14]

# Encoding the categorical feature as factor
# Encoding categorical data
dataset$Geography = as.numeric(factor(dataset$Geography,
                                        levels = c('France', 'Spain', 'Germany'),
                                        labels = c(1, 2, 3)))
dataset$Gender = as.numeric(factor(dataset$Gender,
                                    levels = c('Male', 'Female'),
                                    labels = c(1, 2)))


# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.80)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
training_set[-3] = scale(training_set[-3])
test_set[-3] = scale(test_set[-3])


# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-3])

# Making the Confusion Matrix
cm = table(test_set[, 3], y_pred)

# Importing the dataset
dataset = read.csv('Social_Network_Ads.csv')
dataset = dataset[3:5]

# Encoding the target feature as factor
dataset$Purchased = factor(dataset$Purchased, levels = c(0, 1))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
training_set[-11] = scale(training_set[-11])
test_set[-11] = scale(test_set[-11])

#Fitting ANN to the Training Set
install.packages('h2o')
library(h2o)
h2o.init(nthreads = -1)
classifier = h2o.deeplearning( y = 'Exited',
                               training_frame = as.h2o(training_set),
                               activation = 'Rectifier',
                               epochs = 100,
                               hidden = c(6,6),
                               train_samples_per_iteration = -2)


# Predicting the Test set results
prob_pred = h2o.predict(classifier, newdata = as.h2o(test_set[-11]))
y_pred = (prob_pred > 0.5)
y_pred = as.vector(y_pred)

# Making the Confusion Matrix
cm = table(test_set[, 11], y_pred)

File: deep-learning/artificial-nerural-network/ANN_RC.py
# -*- coding: utf-8 -*-
"""
Created on Wed May 29 10:05:05 2019

@author: ryanc
"""

#Artificail Neural Network

#Part 1 data preprocession

# Classification template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

# Encoding categorical data
# Encoding the Independent Variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:,1:]



# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)


#part 2 making the ANN

#importing the keras libraries
import keras
from keras.models import Sequential
from keras.layers import Dense

from tensorflow.python.client import device_lib
print(device_lib.list_local_devices())

# Initialising the ANN
classifier = Sequential()

#Adding the input layer and the first hidden layer
classifier.add(Dense(output_dim = 6 , kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))

#adding the second hidden layer
classifier.add(Dense(output_dim = 6 , init = 'uniform', activation = 'relu'))

#adding the output layer
classifier.add(Dense(output_dim = 1 , init = 'uniform', activation = 'sigmoid'))

#compiling the ANN
classifier.compile(optimizer = 'adam', loss = 'mean_squared_logarithmic_error', metrics = ['cosine_proximity'] )

#Fitting the ANN to the training set
classifier.fit(X_train, y_train, batch_size=100 , epochs = 100)

#part 3 Making the predictions and evaluating
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)



# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)




























